p2panda-group/src/message_scheme/dcgka.rs

```diff
@@ -0,0 +1,1587 @@
+                control_message: input.control_message,
+                direct_messages: input.direct_messages,
+                me_update_secret: Some(output.sender_update_secret.unwrap()),
+            },
+        ))
+    }
+
+    /// Takes a set of users IDs (including us) and creates a new group with those members.
+    ///
+    /// Note that every member ID needs to be unique for this group.
+    ///
+    /// A group is created in three steps: 1. one user calls "create" and broadcasts a control
+    /// message of type "create" (plus direct messages) to the initial members; 2. each member
+    /// processes that message and broadcasts an "ack" control message; 3. each member processes
+    /// the ack from each other member.
+    pub fn create(
+        y: DcgkaState<ID, OP, PKI, DGM, KMG>,
+        initial_members: Vec<ID>,
+        rng: &Rng,
+    ) -> DcgkaOperationResult<ID, OP, PKI, DGM, KMG> {
+        // De-duplicate members.
+        let mut initial_members: Vec<ID> =
+            initial_members.into_iter().fold(Vec::new(), |mut acc, id| {
+                if !acc.contains(&id) {
+                    acc.push(id);
+                }
+                acc
+            });
+
+        // Add ourselves if the user hasn't done it yet.
+        if !initial_members.contains(&y.my_id) {
+            initial_members.push(y.my_id);
+        }
+
+        // The "create" function constructs the "create" control message.
+        let control_message = ControlMessage::Create(CreateMessage {
+            initial_members: initial_members.clone(),
+        });
+
+        // Generate the set of direct messages to send.
+        let (y_ii, direct_messages) = Self::generate_seed(y, &initial_members, rng)?;
+
+        // It then calls process_create to process the control message for this user (as if it had
+        // received the message) before returning.
+        //
+        // process_create returns a tuple after updating the state including an update secret I; we
+        // use these and ignore the rest.
+        Ok((
+            y_ii,
+            OperationOutput {
+                control_message,
+                direct_messages,
+                me_update_secret: None,
+            },
+        ))
+    }
+
+    /// Called by group members when they receive the "create" message.
+    fn process_create(
+        mut y: DcgkaState<ID, OP, PKI, DGM, KMG>,
+        sender: ID,
+        seq: OP,
+        initial_members: Vec<ID>,
+        direct_message: Option<DirectMessage<ID, OP, DGM>>,
+        rng: &Rng,
+    ) -> DcgkaProcessResult<ID, OP, PKI, DGM, KMG> {
+        y.dgm =
+            DGM::create(y.my_id, &initial_members).map_err(|err| DcgkaError::DgmOperation(err))?;
+        Self::process_seed(y, &sender, seq, direct_message, rng)
+    }
+
+    /// Called by group members when they receive the "ack" message.
+    ///
+    /// In this function, `ackID` and `ackOP` are the user id and id of the acknowledged message.
```

Comment:
Nitpick: in some places we write `id` and in `ID` in others. Would be great to stick with one approach throughout.